framework/src/utils/InputParameterWarehouse.C

```diff
@@ -28,32 +28,73 @@ InputParameterWarehouse::~InputParameterWarehouse()
 
-  // Create the actual InputParameters object to store and reference from the objects
+  // Create the actual InputParameters object that will be reference by the objects
   MooseSharedPointer<InputParameters> ptr(new InputParameters(parameters));
 
+  // Set the name parameter to the object being created
+  ptr->set<std::string>("_object_name") = name;
+
+  // The object name defined by the base class name, this method of storing is used for
+  // determining the uniqueness of the name
+  MooseObjectName unique_name(ptr->get<std::string>("_moose_base"), name);
+
+  // Check that the Parameters do not already exist
+  if (_input_parameters[tid].find(unique_name) != _input_parameters[tid].end())
+    mooseError("A '" << unique_name.tag() << "' object already exists with the name '" << unique_name.name() << "'.\n");
+
+  // Store the parameters according to the base name
+  _input_parameters[tid].insert(std::pair<MooseObjectName, MooseSharedPointer<InputParameters> >(unique_name, ptr));
+
+  // Store the object according to the control tags
+  if (ptr->isParamValid("control_tags"))
+  {
+    std::vector<std::string> tags = ptr->get<std::vector<std::string> >("control_tags");
+    for (std::vector<std::string>::const_iterator it = tags.begin(); it != tags.end(); ++it)
+      _input_parameters[tid].insert(std::pair<MooseObjectName, MooseSharedPointer<InputParameters> >(MooseObjectName(*it, name),  ptr));
+  }
+
   // Set the name and tid parameters
-  ptr->addParam<std::string>("name", name, "The complete name of the object");
   ptr->addPrivateParam<THREAD_ID>("_tid", tid);
   ptr->allowCopy(false);  // no more copies allowed
 
-  // Store the object in the warehouse
-  _name_to_shared_pointer[tid].insert(std::pair<std::string, MooseSharedPointer<InputParameters> >(name, ptr));
+  // Return a reference to the InputParameters object
+  return *ptr;
+}
 
-  return *_name_to_shared_pointer[tid][name];
+
+InputParameters &
+InputParameterWarehouse::getInputParameters(const std::string & name, THREAD_ID tid)
+{
+  return getInputParameters(MooseObjectName(name), tid);
+}
+
+InputParameters &
+InputParameterWarehouse::getInputParameters(const std::string & tag, const std::string & name, THREAD_ID tid)
+{
+  return getInputParameters(MooseObjectName(tag, name), tid);
 }
 
 InputParameters &
-InputParameterWarehouse::getInputParameters(const std::string & name, THREAD_ID tid /* =0 */)
+InputParameterWarehouse::getInputParameters(const MooseObjectName & object_name, THREAD_ID tid)
 {
-  std::map<std::string, MooseSharedPointer<InputParameters> >::iterator iter = _name_to_shared_pointer[tid].find(name);
-  if (iter == _name_to_shared_pointer[tid].end())
-    mooseError("Unknown InputParameters object " << name);
 
-  return *(iter->second.get());
+
+  // Storage for the equal range
+  std::pair<InputParameterIterator, InputParameterIterator> ret;
+
+  // Located the object(s)
+  ret = _input_parameters[tid].equal_range(object_name);
```

Comment:
Don't use equal_range here. If you only want to find _a_ key that matches just use multiapp::find(). It'll turn these 10 lines of code into about 3 with the error message and be cleaner too.